index.js

```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
   })
 }
 
+const allowedTokens = ['[', ']', '{}']
+
+function getSerializedKeyTokens (key) {
+  const tokens = key.split(/(\[|]|{})/).filter(Boolean)
+  return tokens.some(token => allowedTokens.includes(token)) ? tokens : []
+}
+
+function validateSerializedKey (tokens, value) {
+  let isValid = true
+  let token = tokens[0]
+
+  // first token can't be '[', ']', or '{}'
+  if (allowedTokens.includes(token)) isValid = false
+  else if (token === '__proto__' || token === 'constructor') {
+    return `${token} is not allowed as a key name`
+  } else {
+    const bracesEndingCount = tokens.filter(t => t === '{}').length
+
+    // only a single {} special ending is allowed
+    if (bracesEndingCount > 1) isValid = false
+    // {}, if present, must be the last token
+    else if (bracesEndingCount > 0 && tokens[tokens.length - 1] !== '{}') isValid = false
+  }
+
+  const tokensLen = tokens.length
```

Comment:
can we skip this whole part, if isValid is set to false?

Reply:
I removed the whole `isValid` approach. Now, I'm promptly returning when a key is invalid.